diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1566 @@
+                  & size(this%buffer,4), size(this%buffer,5)))
+  select type (buff=>this%buffer)
+    type is (real(r8_kind))
+      buff_out = buff
+  end select
+end subroutine
+
+!> Gets integer(i4_kind) buffer data from 5d buffer object
+!> called through this%get_buffer(i4_outputdata)
+subroutine get_5d_int4 (this, buff_out)
+  class(buffer5d), intent(in) :: this !< 5d allocated buffer object
+  integer(i4_kind), allocatable, intent(out)  :: buff_out(:,:,:,:,:) !< output of copied buffer data
+                                                !! must be the same size as the allocated buffer
+  allocate(buff_out(size(this%buffer,1), size(this%buffer,2), size(this%buffer,3), &
+                  & size(this%buffer,4), size(this%buffer,5)))
+  select type (buff=>this%buffer)
+    type is (integer(i4_kind))
+      buff_out = buff
+  end select
+end subroutine
+
+!> Gets integer(i8_kind) buffer data from 5d buffer object
+!> called through this%get_buffer(i8_outputdata)
+subroutine get_5d_int8 (this, buff_out)
+  class(buffer5d), intent(in) :: this !< 5d allocated buffer object
+  integer(i8_kind), allocatable, intent(out)  :: buff_out(:,:,:,:,:) !< output of copied buffer data
+                                                !! must be the same size as the allocated buffer
+  allocate(buff_out(size(this%buffer,1), size(this%buffer,2), size(this%buffer,3), &
+                  & size(this%buffer,4), size(this%buffer)))
+  select type (buff=>this%buffer)
+    type is (integer(i8_kind))
+      buff_out = buff
+  end select
+end subroutine
+
+!> @brief Initializes a buffer to a given fill value
+subroutine initialize_buffer_0d (this, fillval)
+  class(buffer0d), intent(inout) :: this !< scalar buffer object
+  class(*), intent(in) :: fillval !< fill value, must be same type as the allocated buffer in this
+
+  if(.not. allocated(this%buffer)) call mpp_error(FATAL, 'initialize_buffer_0d:' // &
+      'buffer not yet allocated, allocate_buffer() must be called on this object first.')
+  select type(buff => this%buffer)
+  type is(real(r8_kind))
+    select type(fillval)
+    type is(real(r8_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_0d: mismatch between fill and buffer values')
+    end select
+  type is(real(r4_kind))
+    select type(fillval)
+    type is(real(r4_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_0d: mismatch between fill and buffer values')
+    end select
+  type is(integer(i8_kind))
+    select type(fillval)
+    type is(integer(i8_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_0d: mismatch between fill and buffer values')
+    end select
+  type is(integer(i4_kind))
+    select type(fillval)
+    type is(integer(i4_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_0d: mismatch between fill and buffer values')
```

Comment:
Pass in the field name and whatever other information to make this a more meaningful message.